File: src/content/docs/topics/cs/leetcode-150/backtracking/079-word-search.py
def exist(board, word):
    pass  # TODO: implement

def _run_tests():
    board = [['A', 'B', 'C', 'E'], ['S', 'F', 'C', 'S'], ['A', 'D', 'E', 'E']]
    import copy
    assert exist(copy.deepcopy(board), 'ABCCED') == True
    assert exist(copy.deepcopy(board), 'SEE') == True
    assert exist(copy.deepcopy(board), 'ABCB') == False
    assert exist([['A']], 'A') == True
    assert exist([['A']], 'B') == False
    # --- large-input timing ---
    import time as _t
    _board = [['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'],
              ['K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T'],
              ['U', 'V', 'W', 'X', 'Y', 'Z', 'A', 'B', 'C', 'D'],
              ['E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N'],
              ['O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X'],
              ['Y', 'Z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'],
              ['I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R'],
              ['S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z', 'A', 'B'],
              ['C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L'],
              ['M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V']]
    _t0 = _t.perf_counter()
    exist(_board, 'ABCD')
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf exist(10x10 grid, word="ABCD"): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/backtracking/079-word-search.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func exist(_ board: [[String]], _ word: String) -> Bool {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:77ec7f1cae1ded19e7c67ff5f1c801c1d37a71eb6024045152810e450575f081
    expectEqual(Solution().exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "ABCCED"), true, "canonical-true")
    expectEqual(Solution().exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "SEE"), true, "second-true")
    expectEqual(Solution().exist([["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]], "ABCB"), false, "reuse-false")
    expectEqual(Solution().exist([["A"]], "A"), true, "single")
    expectEqual(Solution().exist([["A"]], "B"), false, "missing-single")
    // EXCLUDED_VECTOR empty-board: [[[]],"A"] | The board has at least one row and one column.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
